src/Terrain/MapPerspective.java

```diff
@@ -241,29 +257,65 @@ public void revealFog(Unit scout, GamePath movepath)
   {
     if (null == viewer)
       return;
+    if( !isFogDoR() )
+    {
+      resetFog(false);
+      return;
+    }
     if( !viewer.isEnemy(scout.CO.army) )
     {
+      UnitContext uc = new UnitContext(this, scout);
       for( PathNode node : movepath.getWaypoints() )
       {
-        for( XYCoord coord : Utils.findVisibleLocations(this, scout, node.x, node.y, false) )
-        {
-          revealFog(coord, false);
-        }
-        // We need to do a second pass with piercing vision so we can know whether to reveal the units
-        for( XYCoord coord : Utils.findVisibleLocations(this, scout, node.x, node.y, true) )
-        {
-          revealFog(coord, true);
-        }
+        uc.setCoord(node.GetCoordinates());
+        uc.calculateVision();
+        revealFog(uc, uc.coord);
+      }
+    }
+  }
+
+  /**
+   * Assumes the scout context has pre-calculated its vision.
+   */
+  protected void revealFog(UnitContext scout, XYCoord seeFrom)
+  {
+    int piercingRange = scout.visionPierces ? scout.visionRange : 1;
+    for( XYCoord coord : Utils.findLocationsInRange(this, seeFrom, 0, piercingRange) )
+    {
+      revealFog(coord, true);
+    }
+    if( !scout.visionPierces )
+      for( XYCoord coord : Utils.findLocationsInRange(this, seeFrom, piercingRange, scout.visionRange) )
+      {
+        revealFog(coord, false);
       }
+  }
+  public void flareTarget(XYCoord origin, int radius)
+  {
+    for( XYCoord coord : Utils.findLocationsInRange(this, origin, 0, radius) )
+    {
+      revealFog(coord, true);
+      flaredTiles.add(coord);
     }
   }
-  
-  public void revealFog(XYCoord coord, boolean piercing)
+
+  protected void revealFog(XYCoord coord, boolean piercing)
   {
     MapLocation loc = master.getLocation(coord);
-    isFogged[coord.x][coord.y] = false;
     lastOwnerSeen[coord.x][coord.y] = loc.getOwner();
-    if (piercing && loc.getResident() != null)
-      confirmedVisibles.add(loc.getResident());
+    Unit resident = loc.getResident();
+
+    TerrainType tt = loc.getEnvironment().terrainType;
+    boolean shouldSee = piercing || !tt.isCover(master.game.rules.fogMode);
+    if( null != resident )
+    {
+      if( piercing )
+        confirmedVisibles.add(resident);
+      else if( !resident.model.hidden ) // Non-invisible aircraft reveal cover that can't repair them.
```

Comment:
invisible aircraft do not?

Reply:
Oh, this is going the other way. An aircraft that is seen will also reveal the terrain underneath it, even if it is (otherwise) cover. Makes sense.